### PR TITLE
Submitted models/marble hd2 sensor config 3

### PR DESCRIPTION
**Pull request by bfotheri**
The marble virtual track team has updated their systems vehicles since we submitted their models. We talked with Angela and she said if they had built the vehicles perhaps you guys would consider these additional sensor configurations. Here are the images showing the two vehicles and reflecting this vehicle sensor configuration. I

tmp_1596226782657
tmp_1596226820604

### CODE REVIEW

File: submitted_models/marble_hd2_sensor_config_3/launch/spawner.rb [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        <left_joint>rear_left_wheel_joint</left_joint>
+        <right_joint>front_right_wheel_joint</right_joint>
+        <right_joint>front_middle_right_wheel_joint</right_joint>
+        <right_joint>rear_middle_right_wheel_joint</right_joint>
+        <right_joint>rear_right_wheel_joint</right_joint>
+        <wheel_separation>#{0.525}</wheel_separation>
+        <wheel_radius>0.129</wheel_radius>
+        <topic>/model/#{_name}/cmd_vel_relay</topic>
+        <min_velocity>-1</min_velocity>
+        <max_velocity>1</max_velocity>
```
Comment (adlarkin) [inline submitted_models/marble_hd2_sensor_config_3/launch/spawner.rb:28]: These velocities seem pretty low - are these the correct values, or should they be higher?

File: submitted_models/marble_hd2_sensor_config_3/specifications.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+At this time, we assume that the motion characteristics are the same as the COSTAR husky which has already been modeled.  We were unable to perform additional validation tests due to COVID-19 restrictions.  
+
+### Endurance Characteristics
+This configuration has an endurance of approximately 2 hours.  We plan to carry out the endurance test characterization but have been prevented from doing so due to the coronavirus measures preventing us from visiting the lab space while preparing these models for simulation.  
+
+### Diversions from Physical Hardware of MARBLE HD2
+Computers were installed in the payload area of the husky and these have been roughly modeled (the rail and computer bay are shown in the model.sdf model).  The MARBLE vehicle uses an AMD Ryzen processor (32-core) with 64 GB of RAM.  It has a cooling system as well which makes up part of the black bay seen in the sdf model.  
+
+## X4 Validation and Specification Links
```
Comment (adlarkin) [inline submitted_models/marble_hd2_sensor_config_3/specifications.md:62]:
```suggestion
## X3 Validation and Specification Links
```

File: submitted_models/marble_hd2_sensor_config_3/launch/vehicle_topics.launch [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+        pkg="subt_ros"
+        type="optical_frame_publisher"
+        name="optical_frame_publisher_front">
+        <remap from="input/image" to="front/image_raw" />
+        <remap from="output/image" to="front/optical/image_raw" />
+        <remap from="input/camera_info" to="front/camera_info" />
+        <remap from="output/camera_info" to="front/optical/camera_info" />
+      </node>
+
+    <!--Backward Facing camera -->
```
Comment (adlarkin) [inline submitted_models/marble_hd2_sensor_config_3/launch/vehicle_topics.launch:90]: I don't think this is mentioned in `specifications.md`, would you mind adding it to the sensor description there?

File: submitted_models/marble_hd2_sensor_config_3/launch/spawner.rb (hunk 1 above)
Comment (bfotheri) [reply submitted_models/marble_hd2_sensor_config_3/launch/spawner.rb:28]: Owing to COVID-19 we do not have validation data. I will ask MARBLE if that has changed in the past couple weeks. Those were the values assigned in leu of not having validation data.

Comment (bfotheri) [reply submitted_models/marble_hd2_sensor_config_3/launch/spawner.rb:28]: Owing to COVID-19 we do not have validation data. I will ask MARBLE if that has changed in the past couple weeks. Those were the values assigned in leu of not having validation data.

**Comment by bfotheri** [comment]
1. I was unable to reproduce this error. The depth cloud points on /ROBOT_NAME/front_down_rgbd_camera/depth/points were visualized in RVIZ with no errors or warnings for me. /ROBOT_NAME/front_down/optical/depth and /ROBOT_NAME/front_down/depth were able to be visualized using rqt_image_view with no problems or errors as well. 

2. Good catch, I've added some lights based off the robot models.

3. I was able to reproduce this. It isn't clear yet to me but it appears to be a combination of the multi-wheel approximation of the tread and perhaps the collision model. We will investigate this but for now the controllers we use are able to compensate for this so it's not a huge problem for us.

4. When I uncomment lines 155-166 of the model.sdf to enable the textures and materials of the model, the simulation is unable to find the paths and consequently crashes. Much trial and error has been unable to solve the problem. It may however just be an artifact of my local setup.

5. The necessary plugin has been added to the spawner.rb file. This should work now.

Let me know if you have any further comments/questions in response to these responses.

File: submitted_models/marble_hd2_sensor_config_3/launch/spawner.rb (hunk 1 above)
Comment (adlarkin) [reply submitted_models/marble_hd2_sensor_config_3/launch/spawner.rb:28]: I know that several other teams/model submissions have faced the same validation data issues due to COVID-19. If you'd like, you can increase the magnitude of these velocities up to `4m/s` until validation data is received - just pick whatever value you believe is reasonable.

Comment (angelacmaio) [reply submitted_models/marble_hd2_sensor_config_3/launch/spawner.rb:28]: The HD2 model will continue to have 1 m/s maximum velocity (matching X1 and Husky) until validation can be provided. Matching the speed to the most similar pre-existing model avoids any new model having a speed advantage that has not been validated.

**Comment by adlarkin** [comment]
> 1. I was unable to reproduce this error. The depth cloud points on /ROBOT_NAME/front_down_rgbd_camera/depth/points were visualized in RVIZ with no errors or warnings for me. /ROBOT_NAME/front_down/optical/depth and /ROBOT_NAME/front_down/depth were able to be visualized using rqt_image_view with no problems or errors as well.

It looks like I was trying to view the wrong topic in rViz - I don't see this error anymore either.

> 2. Good catch, I've added some lights based off the robot models.

Lights look good!

> 3. I was able to reproduce this. It isn't clear yet to me but it appears to be a combination of the multi-wheel approximation of the tread and perhaps the collision model. We will investigate this but for now the controllers we use are able to compensate for this so it's not a huge problem for us.

Although this may not be a problem for your team since you're using controllers, I'm hesitant to release a model with this navigation bug for other teams to use. I'll talk with the other people on the SubT team and get back to you regarding the severity of this bug for release.

> 4. When I uncomment lines 155-166 of the model.sdf to enable the textures and materials of the model, the simulation is unable to find the paths and consequently crashes. Much trial and error has been unable to solve the problem. It may however just be an artifact of my local setup.

The reason why textures are not loading properly is because `MARBLE_HD2_SENSOR_CONFIG_3` needs to exist on fuel or in your fuel cache in order for the texture materials to load (but you're model is not on fuel yet since it is still in review). As a temporary workaround, you _should_ be able to create a `~/.ignition/fuel/fuel.ignitionrobotics.org/OpenRobotics/models/MARBLE_HD2_SENSOR_CONFIG_3/1/` directory and place all relevant files there (`model.*`, `thumbnails/`, `materials/`, and `meshes/`). I tried to do this local workaround, but things still were not working for me. I am looking into this and will get back to you.

> 5. The necessary plugin has been added to the spawner.rb file. This should work now.

I just tried dropping breadcrumbs again, but I am still not seeing anything dropped by the robot. Looking at your latest [commit](https://github.com/osrf/subt/pull/516/commits/962f0e836cb4e33ce6885893624e4eea7ebde121), it doesn't look like `spawner.rb` was modified. Perhaps you forgot to push the changes made to `spawner.rb`? Also, just in case you need it, here is an example of how to add the breadcrumbs plugin to `spawner.rb`:

https://github.com/osrf/subt/blob/a1a886b42f913692dfb8fe9d099ef21ef4f22586/submitted_models/costar_husky_sensor_config_2/launch/spawner.rb#L55-L69

Also, once you're done adding breadcrumbs, would you mind adding to `specifications.md` that this model has breadcrumbs, including the maximum number of breadcrumbs that can be dropped?

File: submitted_models/marble_hd2_sensor_config_3/specifications.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+Its weight is approximately 75 lbs (possible, but heavy to heft with a single person). 
+
+### Sensors
+This HD2 with sensor configuration 1 includes the following sensors. The specifications for these instruments are provided below in the [Validation Links](#validation_links) section.
+
+The following specific sensors are declared payloads of this vehicle:
+
+*D435i RGBD Camera, modeled by rgbd_camera plugin
+  -fixed, downward-facing at about 45 degrees (for examining terrain)
+*Ouster 3D Lidar (64 Channel), modeled by gpu_lidar plugin
```
Comment (adlarkin) [inline submitted_models/marble_hd2_sensor_config_3/specifications.md:31]: It looks like you're using two of these (horizontal and vertical). Can you make a note about the two lidars here?

File: submitted_models/marble_hd2_sensor_config_3/specifications.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+The MARBLE HD2 has the following estimated commercial costs:
+* Base vehicle: $10,000
+* Sensor Suite: $14,000
+* Compute/support electronics: $4,000
+* Total: ~ $28,000
+
+Its weight is approximately 75 lbs (possible, but heavy to heft with a single person). 
+
+### Sensors
+This HD2 with sensor configuration 1 includes the following sensors. The specifications for these instruments are provided below in the [Validation Links](#validation_links) section.
```
Comment (adlarkin) [inline submitted_models/marble_hd2_sensor_config_3/specifications.md:25]:
```suggestion
This HD2 with sensor configuration 3 includes the following sensors. The specifications for these instruments are provided below in the [Validation Links](#validation_links) section.
```

File: submitted_models/marble_hd2_sensor_config_3/specifications.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
+<!---This is a Markdown description of a robot model submitted for inclusion in the DARPA Subterranean Challenge Technology Repository -->
+
+# MARBLE HD2 Sensor Config 3
+This specifications.md file is a description and proof of virtual model validation for the MARBLE HD2 with Sensor Configuration 1. This robot may be launched using an ign launch command with the variable name `marble_hd2_sensor_config_3`.
```
[the remark this answers is not in the document]
Comment (bfotheri) [reply submitted_models/marble_hd2_sensor_config_3/specifications.md:4]: Done

File: submitted_models/marble_hd2_sensor_config_3/specifications.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+### Diversions from Physical Hardware of MARBLE HD2
+Computers were installed in the payload area of the husky and these have been roughly modeled (the rail and computer bay are shown in the model.sdf model).  The MARBLE vehicle uses an AMD Ryzen processor (32-core) with 64 GB of RAM.  It has a cooling system as well which makes up part of the black bay seen in the sdf model.  
+
+## Marble HD2 Validation and Specification Links
+* Vehicle Links:
+  * https://clearpathrobotics.com/husky-unmanned-ground-vehicle-robot/
+
+* Sensor specification links:
+  * Trossen ScorpionX MX-64 Robot Turret - https://www.trossenrobotics.com/p/ScorpionX-RX-64-robot-turret.aspx
```
[the remark this answers is not in the document]
Comment (bfotheri) [reply submitted_models/marble_hd2_sensor_config_3/specifications.md:58]: It is not. Let me remove that.

**Comment by bfotheri** [comment]
DEVELOPMENT: In regards to the strange steering I did a little investigation and found that this problem exists in all the configurations so it is likely to do with the fact that we approximate the treads by using four wheels on each side. When these models were made I did not have access to cave circuit worlds and so I tested them in urban environments. It appears that this vehicle drives fine on flat or tilted surfaces, but when on uneven topography, only some of the wheels are in contact with the ground, often an unequal number on the right and left sides which causes are strange turning to the right or left. I know gazebo had a track vehicle plugin but ignition does not. This is kind of a big item so let me know what you guys think.

**Comment by adlarkin** [comment]
Thank you @acschang and @azeey for your review findings/comments.

> When I uncomment lines 155-166 of the model.sdf to enable the textures and materials of the model, the simulation is unable to find the paths and consequently crashes. Much trial and error has been unable to solve the problem. It may however just be an artifact of my local setup.

@bfotheri I was able to talk with someone else on the SubT team to get this resolved. There's currently a bug that can't handle PBR files paths that are relative if the model has not been uploaded to fuel yet (see this [issue](https://github.com/ignitionrobotics/ign-gazebo/issues/286) and this [comment](https://github.com/ignitionrobotics/ign-gazebo/issues/123#issuecomment-675660707)). Once your model is approved, merged, and uploaded to fuel, this should no longer be an issue. However, for the time being, in order to test this model's materials/textures locally, you can do the following temporary workaround:

1. Change `model://MARBLE_HD2_SENSOR_CONFIG_3/` in [lines 155-167 of model.sdf](https://github.com/bfotheri/subt/blob/14aa4b0c6c7df5d75ceef36c17d4272923acdd8d/submitted_models/marble_hd2_sensor_config_3/model.sdf#L155-L167) to use the full path. On my machine, the model files are located in `/home/developer/subt_ws/src/subt/submitted_models/marble_hd2_sensor_config_3`. This means that lines 155-167 in `model.sdf` would now look like this for me:

```
          <pbr>
            <metal>
              <albedo_map>/home/developer/subt_ws/src/subt/submitted_models/marble_hd2_sensor_config_3/materials/textures/HD2_Albedo.png</albedo_map>
              <metalness_map>/home/developer/subt_ws/src/subt/submitted_models/marble_hd2_sensor_config_3/materials/textures/HD2_Metalness.png</metalness_map>
              <roughness_map>/home/developer/subt_ws/src/subt/submitted_models/marble_hd2_sensor_config_3/materials/textures/HD2_Roughness.png</roughness_map>
            </metal>
          </pbr>
          <!-- fallback to script if no PBR support-->
          <script>
            <uri>/home/developer/subt_ws/src/subt/submitted_models/marble_hd2_sensor_config_3/materials/scripts/</uri>
            <uri>/home/developer/subt_ws/src/subt/submitted_models/marble_hd2_sensor_config_3/materials/textures/</uri>
            <name>HD/HD2_Diffuse</name>
          </script>
```

2. Create a fuel cache for this model, which is done by creating a `~/.ignition/fuel/fuel.ignitionrobotics.org/OpenRobotics/models/MARBLE_HD2_SENSOR_CONFIG_3/1/` directory:

```
$ mkdir -p ~/.ignition/fuel/fuel.ignitionrobotics.org/OpenRobotics/models/MARBLE_HD2_SENSOR_CONFIG_3/1/
```

3. Copy `materials/`, `meshes/`, `thumbnails/`, and `model.*` over to the newly created fuel cache directory:

```
# go to the model's directory in the subt repo
# (this path will probably be different on your machine)
$ cd /home/developer/subt_ws/src/subt/submitted_models/marble_hd2_sensor_config_3

$ cp -r materials/ ~/.ignition/fuel/fuel.ignitionrobotics.org/OpenRobotics/models/MARBLE_HD2_SENSOR_CONFIG_3/1/
$ cp -r meshes/ ~/.ignition/fuel/fuel.ignitionrobotics.org/OpenRobotics/models/MARBLE_HD2_SENSOR_CONFIG_3/1/
$ cp -r thumbnails/ ~/.ignition/fuel/fuel.ignitionrobotics.org/OpenRobotics/models/MARBLE_HD2_SENSOR_CONFIG_3/1/
$ cp model.* ~/.ignition/fuel/fuel.ignitionrobotics.org/OpenRobotics/models/MARBLE_HD2_SENSOR_CONFIG_3/1/
```

4. Build the workspace with the `subt` repo and start a simulator with `MARBLE_HD2_SENSOR_CONFIG_3`. Here is how the model looks for me with the materials/textures loaded:

marble_model_with_materials_textures

As @acschang [mentioned](https://github.com/osrf/subt/pull/516#pullrequestreview-472886451), it looks like some sensors are "floating" in this model.

File: submitted_models/marble_hd2_sensor_config_3/model.sdf [hunk 8]
```diff
@@ -0,0 +1,1181 @@
...
+          <iyz>0</iyz>
+          <izz>0.02467</izz>
+        </inertia>
+      </inertial>
+      <collision name="rear_right_wheel_link_collision">
+        <pose frame="">0 0 0 1.5707963267948966 -0 0</pose>
+        <geometry>
+          <cylinder>
+            <length>0.1143</length>
+            <radius>0.1651</radius>
```
[the remark this answers is not in the document]
Comment (bfotheri) [reply submitted_models/marble_hd2_sensor_config_3/model.sdf:1141]: It is not. That may be why it is driving strangely. Good catch. Thank you.

**Comment by bfotheri** [comment]
Update: The driving issue has been fixed. One of the wheels was larger than the others. This has been rectified in my latest commits and the behavior appears to have been resolved. @azeey Thank you for finding this!

**Review by acschang — COMMENTED** [review]
This sensor configuration is under consideration for the final circuit.  The currently outstanding tasks are listed below:

- Please connect floating sensors and add shading to the existing robot mesh. We will provide assistance in removing preexisting sensors that are not present on this robot model sensor configuration from the single mesh if you are unable to.
- The provided documentation for the HD MIPI cameras lists the cameras to have a 808x608 resolution, 50 Hz frame rate, and the listed lens has a 80 degree FoV.  The parameters listed for the modeled sensor are 1280x960 resolution, 15 Hz frame rate, and a 60 degree FoV.  Please either correct the discrepancy or provide an explanation for the simulated sensor parameters differing from the provided documentation.
- The validation data is missing. Please complete the validation testing as required by the [Simulation Model Preparation Guide](https://subtchallenge.com/resources/Simulation_Model_Preparation_Guide.pdf) and submit the required data.

**Comment by nkoenig** [comment]
The bounding box for this model is
```
Min[-0.689611 -0.332041 -0.135] Max[0.5479 0.332197 0.68003]
```

**Pull request closed without merging**